capture/src/v0_endpoint.rs

```diff
@@ -86,41 +89,36 @@ pub async fn event(
                     tracing::error!("failed to decode form data: {}", e);
                     CaptureError::RequestDecodingError(String::from("missing data field"))
                 })?;
-            RawEvent::from_bytes(payload.into())
+            RawRequest::from_bytes(payload.into())
         }
         ct => {
             tracing::Span::current().record("content_type", ct);
 
-            RawEvent::from_bytes(body)
+            RawRequest::from_bytes(body)
         }
     }?;
 
+    let sent_at = request.sent_at().or(meta.sent_at());
+    let token = match request.extract_and_verify_token() {
+        Ok(token) => token,
+        Err(err) => {
+            report_dropped_events("token_shape_invalid", request.events().len() as u64);
+            return Err(err);
+        }
+    };
+    let is_historical = request.is_historical(); // TODO: use to write to historical topic
+    let events = request.events(); // Takes ownership of request
+
+    tracing::Span::current().record("token", &token);
+    tracing::Span::current().record("is_historical", is_historical);
     tracing::Span::current().record("batch_size", events.len());
 
     if events.is_empty() {
         return Err(CaptureError::EmptyBatch);
     }
```

Comment:
Funnily enough, currently when sending an empty batch to `/batch` I get a 400 with the following error detail:
```
"Invalid payload: All events must have the event name field \"event\"!"
```
Which seems to be better represented here by the `CaptureError::MissingEventName` variant instead of the `EmptyBatch` variant. I assume django is trying to validate each event in batch immediately. Anyways, I do think capture.py is wrong here and `EmptyBatch` should be be the correct error.

Reply:
On the `/batch` shape (the `RawRequest::Batch` enum variant), the token is taken from the mandatory toplevel field. The code to scan all events is for the `RawRequest::Array` shape (that's the posthog-js shape with an array of events and no toplevel metadata).

python will most probably fallback to scanning events if the toplevel token is not given, but the double-run will tell us whether this fallback is needed.